### PR TITLE
detect/detection_filter: extend unique_on with src_ip|dst_ip - v7

### PR DESCRIPTION
**Pull request by oferda4**
Add optional unique_on {src_ip|dst_ip} to detection_filter for distinct IP address counting within the seconds window. Features:

Runtime uses a hash table per threshold entry for tracking unique IP addresses (both IPv4 and IPv6).
Follows detection_filter semantics: alerting starts after the threshold (>count), not at it.
On window expiry, the window is reset and the current packet's IP is recorded as the first distinct of the new window. Validation:
unique_on src_ip/dst_ip works with any IP protocol (unlike port-based unique_on which requires tcp/udp/sctp). Memory management:
Hash table memory is bounded by detect.thresholds.memcap.
Reuses existing counters: bitmap_memuse and bitmap_alloc_fail. Refactoring:
Added ThresholdDistinctAdd helper to consolidate port/IP tracking logic and reduce code duplication. Tests:
C unit tests for parsing unique_on src_ip and dst_ip options.

Changes:
v2:
- Update documentation
- Increase unit tests coverage
- Document use of hash table

v3:
- Cleanup unit tests

v4:
- Added validation to reject duplicate unique_on options
- Fixed memory accounting in ThresholdDistinctReset: counters are now properly decremented if hash table re-init fails
- Hash table is now pre-sized to count buckets instead of a fixed 64

v5:
- Fixed use-after-free: HashTableAdd stores pointers without copying, so ThresholdDistinctAddIP now allocates owned copies of IP address bytes. A Free callback (ThresholdDistinctIPFree) is registered with the hash table to automatically free copies on reset/free.
- Handle HashTableAdd allocation failure by freeing the IP copy
- Simplified window expiry: te->current_count = 1 is now unconditional
- Added documentation on two-phase memory allocation model (init-time bucket array vs. runtime per-entry copies)

v6:
- Replace HashTable with pre-allocated flat buffer (Address *distinct_ip_buf) for IP deduplication — eliminates all per-packet SCMalloc calls
- Add 4 new unit tests: IPv4 basic + buffer-full boundary, duplicate suppression, window reset, dst_ip with track by_src
- Restrict port-protocol validation to unique_on src_port/dst_port only (not src_ip/dst_ip)

v7:
- Simplified `else if` to `else` in ThresholdDistinctInit (only remaining case after port branch)                                                             
- Clarified scan_limit comment: current_count can exceed max_count (counts all distinct IPs seen, but only max_count are stored in the buffer)                
- Added DF_UNIQUE_IP_MAX_COUNT (10000) cap for IP-based unique_on to bound per-entry buffer size                                                              

Previous PR: https://github.com/OISF/suricata/pull/15164

Ticket: https://redmine.openinfosecfoundation.org/issues/8250
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2924

### CODE REVIEW

File: src/detect-engine-threshold.c [hunk 1]
```diff
@@ -240,6 +276,65 @@ static inline void ThresholdDistinctAddPort(ThresholdEntry *te, uint16_t port)
...
+        COPY_ADDRESS(addr, &te->distinct_ip_buf[te->current_count]);
+    }
+    te->current_count++;
+}
+
+/**
+ * \brief Try to add a distinct value (port or IP) based on unique_on type.
+ * \return true if distinct tracking is active (bitmap/hash allocated), false if fallback needed
+ */
+static inline bool ThresholdDistinctAdd(
```
Comment (catenacyber) [inline src/detect-engine-threshold.c:319]: nit : I still find it a bit weird that this returns a bool value.

Could we not return anything and do the `te->current_count++;` in case the init alloc failed ?

File: src/detect-threshold.h [hunk 2]
```diff
@@ -48,10 +48,15 @@
 #define TH_ACTION_REJECT    0x20
 
 /* distinct counting support (for detection_filter) */
+#define DF_UNIQUE_IP_MAX_COUNT                                                                     \
```
Comment (catenacyber) [inline src/detect-threshold.h:51]: nit: should this be in the only C file using it ?

Comment (catenacyber) [reply src/detect-threshold.h:51]: Also wondering if we should limit to 1024...

**Review by catenacyber — APPROVED** [review]
Thanks for the work, I think it is good enough

CI : ✅  (unrelated failure)
Code : good enough even if I made some nits
Commits segmentation : cool thanks
Commit messages : good thanks
Git ID set : ok
CLA : you already contributed
Doc update : cool
Redmine ticket : ok
Rustfmt : no rust
Tests : looks good
Dependencies added: none

**Comment by oferda4** [comment]
@catenacyber thank you for the approve. I'll make a new PR version to make the fixes in a clean way
